src/util-detect-file-hash.c

```diff
@@ -0,0 +1,377 @@
+        char *str, uint8_t type)
+{
+    DetectFileHashData *filehash = NULL;
+    FILE *fp = NULL;
+    char *filename = NULL;
+
+    /* We have a correct filemd5 option */
+    filehash = SCMalloc(sizeof(DetectFileHashData));
+    if (unlikely(filehash == NULL))
+        goto error;
+
+    memset(filehash, 0x00, sizeof(DetectFileHashData));
+
+    if (strlen(str) && str[0] == '!') {
+        filehash->negated = 1;
+        str++;
+    }
+
+    if (type == DETECT_FILEMD5) {
+        filehash->hash = ROHashInit(18, 16);
+    }
+    else if (type == DETECT_FILESHA1) {
+        filehash->hash = ROHashInit(18, 20);
+    }
+    else if (type == DETECT_FILESHA256) {
+        filehash->hash = ROHashInit(18, 32);
+    }
+
+    if (filehash->hash == NULL) {
+        goto error;
+    }
+
+    /* get full filename */
+    filename = DetectLoadCompleteSigPath(de_ctx, str);
+    if (filename == NULL) {
+        goto error;
+    }
+
+    char line[8192] = "";
+    fp = fopen(filename, "r");
+    if (fp == NULL) {
+        SCLogError(SC_ERR_OPENING_RULE_FILE, "opening hash file %s: %s", filename, strerror(errno));
+        goto error;
+    }
+
+    int line_no = 0;
+    while(fgets(line, (int)sizeof(line), fp) != NULL) {
+        size_t len = strlen(line);
+        line_no++;
+
+        /* ignore comments and empty lines */
+        if (line[0] == '\n' || line [0] == '\r' || line[0] == ' ' || line[0] == '#' || line[0] == '\t')
+            continue;
+
+        while (isspace(line[--len]));
+
+        /* Check if we have a trailing newline, and remove it */
+        len = strlen(line);
+        if (len > 0 && (line[len - 1] == '\n' || line[len - 1] == '\r')) {
+            line[len - 1] = '\0';
+        }
+
+        /* cut off longer lines than a SHA256 represented in hexadecimal  */
+        if (strlen(line) > 64)
+            line[64] = 0x00;
+
+        if (LoadHashTable(filehash->hash, line, filename, line_no, type) != 1) {
+            goto error;
+        }
+    }
+    fclose(fp);
+    fp = NULL;
+
+    if (ROHashInitFinalize(filehash->hash) != 1) {
+        goto error;
+    }
+    SCLogInfo("Hash hash table size %u bytes%s", ROHashMemorySize(filehash->hash), filehash->negated ? ", negated match" : "");
+
+    SCFree(filename);
+    return filehash;
+
+error:
+    if (filehash != NULL)
+        DetectFileHashFree(filehash);
+    if (fp != NULL)
+        fclose(fp);
+    if (filename != NULL)
+        SCFree(filename);
+    return NULL;
+}
+
+/**
+ * \brief this function is used to parse filemd5, filesha1 and filesha256 options
+ * \brief into the current signature
+ *
+ * \param de_ctx pointer to the Detection Engine Context
+ * \param s pointer to the Current Signature
+ * \param str pointer to the user provided "filemd5", "filesha1" or "filesha256" option
+ * \param type type of file hash to setup
+ *
+ * \retval 0 on Success
+ * \retval -1 on Failure
+ */
+int DetectFileHashSetup (DetectEngineCtx *de_ctx, Signature *s, char *str,
+        uint8_t type)
```

Comment:
probably safer to use 'int' here, as the number of supported keywords is growing. Would be a hard to find bug later ;)